# BBtools #
## A suite of command line utilities for sequence quality control, mapping, assembly, and kmer based analysis. ##
### Written by Brian Bushnell. ###
These examples are maintained by Adam Rivers.

As the BBtools website describes it:

> BBTools is a suite of fast, multithreaded bioinformatics tools designed for analysis of DNA and RNA sequence data. BBTools can handle common sequencing file formats such as fastq, fasta, sam, scarf, fasta+qual, compressed or raw, with autodetection of quality encoding and interleaving. It is written in Java and works on any platform supporting Java, including Linux, MacOS, and Microsoft Windows and Linux; there are no dependencies other than Java (version 7 or higher). Program descriptions and options are shown when running the shell scripts with no parameters.

> BBTools is open source and free for unlimited use, and is used regularly by DOE JGI and other institutions around the world.

> The BBTools suite includes programs such as:
> * bbduk – filters or trims reads for adapters and contaminants using k-mers
> * bbmap – short-read aligner for DNA and RNA-seq data
> * bbmerge – merges overlapping or nonoverlapping pairs into a single reads
> * reformat – converts sequence files between different formats such as fastq and fasta

For more information see the [BBtools website](http://jgi.doe.gov/data-and-tools/bbtools/) and the [BBtools user guide](http://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide). Support is pavailable from a number of threads at the [SeqAnswers](http://seqanswers.com/forums/showthread.php?t=41057) site. A listing of those threads is provided [here](http://jgi.doe.gov/data-and-tools/bbtools/bbtools-faq-support-forums/).

The package contains 125 scripts and programs but this examples folder highlights some of the most common use cases.

___
## Contents ##
1. `bbmap_examples` - A directory with bbmap examples
  * `basic_mapping` - A directory with a shell script that generates an index then maps reads to the index
    * `bbmap_basic_mapping.sh` - example shell script
    * `results` - Data files generated by this analysis will be placed in this directory. This format is used in other examples.
2. `bbduk_examples` - A directory with examples of BBDuk
  * `adaptor_trimming` - A directory with a shell script that trims sequencing adaptors of a sequence fastq
    * `bbduk_adaptor_trimming.sh` - example shell script
  * `quality_trimming` - A directory with a shell script that trims low quality bases
    * `bbduk_quality_trimming.sh` - example shell script
  * `kmer filtering` - A directory with a shell script that removes reads containing kmers from a reference. Often used to remove PhiX spike in sequences that have made it past Illumina barcode based removal.
    * `bbduk_kmer_filtering.sh` - example shell script
  * `histogram_generation` - A directory with a shell script that generates quality histograms
    * `bbduk_histogram_generation.sh` - example shell script
3. `bbmerge_examples` - A directory containing examples for the multithreaded read merging tool bbmerge
  * `basic_merging` -  A directory with a shell script that merges overlapping reads and creates insert size histogram
    *  `bbmerge_basic_merging.sh` - example shell script
  * `error_correction` -  A directory with a shell script that corrects errors in sequencing but retuns the unmerged reads
    * `bbmerge_error_correction.sh` - example shell scripts

### Running notes ###
* For simplicity I have created an env.sh variable that creates the top level path directory ($bbtoolsExamplesDir). By sourcing this file ("source env.sh") you can run the example scripts from different directories.
* `setup.sh` is a maintenance script that can be used with the `build` or `clean` option to run all examples and place the data into the results directories.
* If an example script is run individually progress and statistics will be written to the screen through the standard error stream. That output can be seen in the `terminaloutput.txt`  file of the `results` directory.
* To test this interactively on CERES log into an interactive node e.g.  `srun --pty -p short -t 01:00:00 -n 16 -N 1 /bin/bash -I`
